### PR TITLE
feat: `avm/res/db-for-my-sql/flexible-server`

### PR DESCRIPTION
**Pull request by segraef**
[avm.res.db-for-my-sql.flexible-server](https://github.com/segraef/bicep-registry-modules/actions/workflows/avm.res.db-for-my-sql.flexible-server.yml)

### CODE REVIEW

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]
Requires uncommenting in the `/.github/CODEOWNERS` file

**Comment by segraef** [comment]
> Requires uncommenting in the `/.github/CODEOWNERS` file

done

File: avm/res/db-for-my-sql/flexible-server/main.bicep [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+@description('Optional. The firewall rules to create in the MySQL flexible server.')
+param firewallRules array = []
+
+@description('Optional. Array of role assignment objects that contain the "roleDefinitionIdOrName" and "principalId" to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: "/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11".')
+param roleAssignments roleAssignmentType
+
+@description('Optional. The diagnostic settings of the service.')
+param diagnosticSettings diagnosticSettingType
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
```
Comment (AlexanderSehr) [inline avm/res/db-for-my-sql/flexible-server/main.bicep:162]:
```suggestion
@description('Optional. Enable/Disable usage telemetry for module.')
```

File: avm/res/db-for-my-sql/flexible-server/main.bicep [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+@description('Optional. MySQL Server version.')
+param version string = '5.7'
+
+@description('Optional. The databases to create in the server.')
+param databases array = []
+
+@description('Optional. The firewall rules to create in the MySQL flexible server.')
+param firewallRules array = []
+
+@description('Optional. Array of role assignment objects that contain the "roleDefinitionIdOrName" and "principalId" to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: "/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11".')
```
Comment (AlexanderSehr) [inline avm/res/db-for-my-sql/flexible-server/main.bicep:156]:
```suggestion
@description('Optional. Array of role assignments to create.')
```

File: avm/res/db-for-my-sql/flexible-server/main.bicep [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+  @description('Optional. The name of logs that will be streamed. "allLogs" includes all possible logs for the resource. Set to \'\' to disable log collection.')
+  logCategoriesAndGroups: {
+    @description('Optional. Name of a Diagnostic Log category for a resource type this setting is applied to. Set the specific logs to collect here.')
+    category: string?
+
+    @description('Optional. Name of a Diagnostic Log category group for a resource type this setting is applied to. Set to `allLogs` to collect all logs.')
+    categoryGroup: string?
+  }[]?
+
+  @description('Optional. The name of logs that will be streamed. "allMetrics" includes all possible logs for the resource. Set to \'\' to disable log collection.')
```
Comment (AlexanderSehr) [inline avm/res/db-for-my-sql/flexible-server/main.bicep:424]:
```suggestion
  @description('Optional. The name of metrics that will be streamed. "allMetrics" includes all possible metrics for the resource. Set to \'\' to disable metric collection.')
```

File: .github/CODEOWNERS [hunk 4]
```diff
@@ -38,7 +38,7 @@
 #/avm/res/databricks/workspace/ @Azure/avm-res-databricks-workspace-module-owners-bicep @Azure/avm-core-team-technical-bicep
 /avm/res/data-factory/factory/ @Azure/avm-res-datafactory-factory-module-owners-bicep @Azure/avm-core-team-technical-bicep
 #/avm/res/data-protection/backup-vault/ @Azure/avm-res-dataprotection-backupvault-module-owners-bicep @Azure/avm-core-team-technical-bicep
-#/avm/res/db-for-my-sql/flexible-server/ @Azure/avm-res-dbformysql-flexibleserver-module-owners-bicep @Azure/avm-core-team-technical-bicep
+/avm/res/db-for-my-sql/flexible-server/ @Azure/avm-res-dbformysql-flexibleserver-module-owners-bicep @Azure/avm-core-team-technical-bicep
```
Comment (eriqua) [inline .github/CODEOWNERS:41]: Please ref to https://azure.github.io/Azure-Verified-Modules/specs/shared/#team-memberships

File: .github/workflows/avm.res.db-for-my-sql.flexible-server.yml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+concurrency:
+  group: ${{ github.workflow }}
+
+jobs:
+  ###########################
+  #   Initialize pipeline   #
+  ###########################
+  job_initialize_pipeline:
+    runs-on: ubuntu-20.04
```
Comment (eriqua) [inline .github/workflows/avm.res.db-for-my-sql.flexible-server.yml:47]:
```suggestion
    runs-on: ubuntu-latest
```

File: .github/CODEOWNERS (hunk 4 above)
Comment (segraef) [reply .github/CODEOWNERS:41]: @Azure/avm-res-dbformysql-flexibleserver-module-owners-bicep
@Azure/avm-res-dbformysql-flexibleserver-module-contributors-bicep

File: avm/res/db-for-my-sql/flexible-server/tests/e2e/max/main.test.bicep [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+// ========== //
+// Parameters //
+// ========== //
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'dep-${namePrefix}-dbformysql.flexibleservers-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = 'eastus'
```
Comment (AlexanderSehr) [inline avm/res/db-for-my-sql/flexible-server/tests/e2e/max/main.test.bicep:15]: Will be overwritten. Please refer to: https://github.com/Azure/bicep-registry-modules/pull/792#discussion_r1455072261

File: avm/res/db-for-my-sql/flexible-server/tests/e2e/max/main.test.bicep [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    lock: {
+      kind: 'CanNotDelete'
+      name: 'myCustomLockName'
+    }
+    roleAssignments: [
+      {
+        roleDefinitionIdOrName: 'Reader'
+        principalId: nestedDependencies2.outputs.managedIdentityPrincipalId
+        principalType: 'ServicePrincipal'
+      }
```
Comment (AlexanderSehr) [inline avm/res/db-for-my-sql/flexible-server/tests/e2e/max/main.test.bicep:100]:
```suggestion
      {
        roleDefinitionIdOrName: 'Owner'
        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
        principalType: 'ServicePrincipal'
      }
      {
        roleDefinitionIdOrName: 'b24988ac-6180-42a0-ab88-20f7382dd24c'
        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
        principalType: 'ServicePrincipal'
      }
      {
        roleDefinitionIdOrName: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
        principalType: 'ServicePrincipal'
      }
```

For reference. Requires regeneration of readme.

File: avm/res/db-for-my-sql/flexible-server/tests/e2e/private/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-diagnosticDependencies'
+  params: {
+    storageAccountName: 'dep${namePrefix}diasa${serviceShort}01'
+    logAnalyticsWorkspaceName: 'dep-${namePrefix}-law-${serviceShort}'
+    eventHubNamespaceEventHubName: 'dep-${namePrefix}-evh-${serviceShort}'
+    eventHubNamespaceName: 'dep-${namePrefix}-evhns-${serviceShort}'
+    location: location
+  }
+}
```
Comment (AlexanderSehr) [inline avm/res/db-for-my-sql/flexible-server/tests/e2e/private/main.test.bicep:60]:
```suggestion
```

I'd argue it's not relevant for the purpose of this test / illustration of how a private server can be deployed

File: avm/res/db-for-my-sql/flexible-server/tests/e2e/private/main.test.bicep [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        roleDefinitionIdOrName: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
+        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
+        principalType: 'ServicePrincipal'
+      }
+    ]
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      resourceType: 'MySQL Flexible Server'
+      serverName: '${namePrefix}${serviceShort}001'
+    }
```
Comment (AlexanderSehr) [inline avm/res/db-for-my-sql/flexible-server/tests/e2e/private/main.test.bicep:98]:
```suggestion
```

I'd argue it's not relevant for the purpose of this test / illustration of how a private server can be deployed

File: avm/res/db-for-my-sql/flexible-server/tests/e2e/private/main.test.bicep [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+          {
+            category: 'AllMetrics'
+          }
+        ]
+        eventHubName: diagnosticDependencies.outputs.eventHubNamespaceEventHubName
+        eventHubAuthorizationRuleResourceId: diagnosticDependencies.outputs.eventHubAuthorizationRuleId
+        storageAccountResourceId: diagnosticDependencies.outputs.storageAccountResourceId
+        workspaceResourceId: diagnosticDependencies.outputs.logAnalyticsWorkspaceResourceId
+      }
+    ]
```
Comment (AlexanderSehr) [inline avm/res/db-for-my-sql/flexible-server/tests/e2e/private/main.test.bicep:135]:
```suggestion
```

I'd argue it's not relevant for the purpose of this test / illustration of how a private server can be deployed

File: avm/res/db-for-my-sql/flexible-server/tests/e2e/private/main.test.bicep (hunk 9 above)
Comment (AlexanderSehr) [inline avm/res/db-for-my-sql/flexible-server/tests/e2e/private/main.test.bicep:98]:
```suggestion
```

I'd argue it's not relevant for the purpose of this test / illustration of how a private server can be deployed

File: avm/res/db-for-my-sql/flexible-server/README.md [hunk 11]
```diff
@@ -0,0 +1,1381 @@
...
+      kind: 'CanNotDelete'
+      name: 'myCustomLockName'
+    }
+    managedIdentities: {
+      userAssignedResourceIds: [
+        '<geoBackupManagedIdentityResourceId>'
+        '<managedIdentityResourceId>'
+      ]
+    }
+    roleAssignments: [
```
Comment (AlexanderSehr) [inline avm/res/db-for-my-sql/flexible-server/README.md:194]: Seems outdated, please refresh given the test file updates

Comment (segraef) [reply avm/res/db-for-my-sql/flexible-server/README.md:194]: done

File: avm/res/db-for-my-sql/flexible-server/tests/e2e/max/main.test.bicep (hunk 6 above)
Comment (segraef) [reply avm/res/db-for-my-sql/flexible-server/tests/e2e/max/main.test.bicep:15]: set tempLocation

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**